utils/frame/try-runtime/cli/src/commands/follow_chain.rs

```diff
@@ -42,51 +46,177 @@ pub struct FollowChainCmd {
+	let client = WsClientBuilder::default()
+		.connection_timeout(std::time::Duration::new(20, 0))
+		.max_notifs_per_subscription(1024)
+		.max_request_body_size(u32::MAX)
+		.build(url)
+		.await
+		.unwrap();
+
+	log::info!(target: LOG_TARGET, "subscribing to {:?} / {:?}", SUB, UN_SUB);
+
+	let sub = client.subscribe(SUB, None, UN_SUB).await.unwrap();
+	(client, sub)
+}
+
+/// Abstraction over RPC calling for headers.
+#[async_trait]
+trait HeaderProvider<Block: BlockT>
+where
+	Block::Header: HeaderT,
+{
+	/// Awaits for the header of the block with hash `hash`.
+	async fn get_header(&mut self, hash: Block::Hash) -> Block::Header;
+}
+
+struct RpcHeaderProvider<Block: BlockT> {
+	uri: String,
+	_phantom: PhantomData<Block>,
+}
+
+#[async_trait]
+impl<Block: BlockT> HeaderProvider<Block> for RpcHeaderProvider<Block>
+where
+	Block::Header: DeserializeOwned,
+{
+	async fn get_header(&mut self, hash: Block::Hash) -> Block::Header {
+		rpc_api::get_header::<Block, _>(&self.uri, hash).await.unwrap()
+	}
+}
+
+/// Abstraction over RPC subscription for finalized headers.
+#[async_trait]
+trait HeaderSubscription<Block: BlockT>
+where
+	Block::Header: HeaderT,
+{
+	/// Await for the next finalized header from the subscription.
+	///
+	/// Returns `None` if either the subscription has been closed or there was an error when reading
+	/// an object from the client.
+	async fn next_header(&mut self) -> Option<Block::Header>;
+}
+
+#[async_trait]
+impl<Block: BlockT> HeaderSubscription<Block> for Subscription<Block::Header>
+where
+	Block::Header: DeserializeOwned,
+{
+	async fn next_header(&mut self) -> Option<Block::Header> {
+		match self.next().await {
+			Some(Ok(header)) => Some(header),
+			None => {
+				log::warn!("subscription closed");
+				None
+			},
+			Some(Err(why)) => {
+				log::warn!("subscription returned error: {:?}. Probably decoding has failed.", why);
+				None
+			},
+		}
+	}
+}
+
+/// Stream of all finalized headers.
+///
+/// Returned headers are guaranteed to be ordered. There are no missing headers (even if some of
+/// them lack justification).
+struct FinalizedHeaders<Block: BlockT, HP: HeaderProvider<Block>, HS: HeaderSubscription<Block>> {
+	header_provider: HP,
+	subscription: HS,
+	fetched_headers: VecDeque<Block::Header>,
+	last_returned: Option<<Block::Header as HeaderT>::Number>,
+}
+
+impl<Block: BlockT, HP: HeaderProvider<Block>, HS: HeaderSubscription<Block>>
+	FinalizedHeaders<Block, HP, HS>
+where
+	<Block as BlockT>::Header: DeserializeOwned,
+{
+	pub fn new(header_provider: HP, subscription: HS) -> Self {
+		Self {
+			header_provider,
+			subscription,
+			fetched_headers: VecDeque::new(),
+			last_returned: None,
+		}
+	}
+
+	/// Reads next finalized header from the subscription. If some headers (without justification)
+	/// have been skipped, fetches them as well.
+	///
+	/// All fetched headers are stored in `self.fetched_headers`.
+	async fn fetch(&mut self) {
```

Comment:
This could return number of headers that it updated in `self.fetched_headers`.

Reply:
sure, done